08-trees/SearchTree/dictionary.c

```diff
@@ -0,0 +1,201 @@
+#include <stdio.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "dictionary.h"
+
+typedef struct {
+    char* value;
+    int key;
+} NodeValue;
+
+typedef struct Node {
+    NodeValue value;
+    Node* left;
+    Node* right;
+} Node;
+
+Node* createNode(NodeValue value, bool* errorCode) {
+    Node* node = calloc(1, sizeof(Node));
+    if (node == NULL) {
+        *errorCode = true;
+        return NULL;
+    }
+    node->value = value;
+    return node;
+}
+
+Node* createTree(const int key, const char* value, bool* errorCode) {
+    Node* root = calloc(1, sizeof(Node));
+    if (root == NULL) {
+        *errorCode = true;
+        return NULL;
+    }
+    root->value.key = key;
+
+    const char* copyOfValue = strdup(value);
+    if (copyOfValue == NULL) {
+        free(root);
+        *errorCode = true;
+        return NULL;
+    }
+    root->value.value = copyOfValue;
+    return root;
+}
+
+void deleteTree(Node** root) {
+    if (*root == NULL) {
+        return;
+    }
+    deleteTree((&(*root)->left));
+    deleteTree((&(*root)->right));
+
+    if ((*root)->value.value != NULL) {
+        free((*root)->value.value);
+    }
+
+    free(*root);
+    *root = NULL;
+}
+
+Node* addNode(Node* node, const int key, const char* value, bool* errorCode) {
+    if (key == node->value.key) {
+        node->value.value = value;
+        return node;
+    }
+
+    if (node->left == NULL && key < node->value.key) {
+        Node* newNode = createTree(key, value, errorCode);
+        if (*errorCode) {
+            return node;
+        }
+        node->left = newNode;
+        return node;
+    }
+    else if (node->right == NULL && key > node->value.key) {
+        Node* newNode = createTree(key, value, errorCode);
+        if (*errorCode) {
+            return node;
+        }
+        node->right = newNode;
+        return node;
+    }
+
+    if (key > node->value.key) {
+        addNode(node->right, key, value, errorCode);
+        return node;
+    }
+    if (key < node->value.key) {
+        addNode(node->left, key, value, errorCode);
+        return node;
+    }
+}
+
+char* searchByKey(Node* node, const int key) {
+    if (node == NULL) {
+        return NULL;
+    }
+
+    if (node->value.key == key) {
+        return node->value.value;
+    }
+
+    if ((node->left == NULL && node->value.key > key) || (node->right == NULL && node->value.key < key)) {
+        return NULL;
+    }
+
+    const char* foundValue = NULL;
+    if (node->value.key < key) {
+        foundValue = searchByKey(node->right, key);
+    }
+    if (node->value.key > key) {
+        foundValue = searchByKey(node->left, key);
+    }
+    return foundValue;
```

Comment:
```suggestion
    if (node->value.key < key) {
        return searchByKey(node->right, key);
    }
    return searchByKey(node->left, key);
```
Или даже так:
```c
    return node->value.key < key ? searchByKey(node->right, key) : searchByKey(node->left, key);
```
Или так:
```c
    return searchByKey(node->value.key < key ? node->right : node->left, key);
```